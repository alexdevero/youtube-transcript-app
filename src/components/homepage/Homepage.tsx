'use client'

import { type FC, useState } from 'react'

import { Button } from '@/components/button/Button'
import { Input } from '@/components/input/Input'
import { useGetYTTranscription } from '@/hooks/use-get-yt-transcription'
import { isValidYoutubeVideoUrl } from '@/utils/validators'

const saveToFile = (data: string, filename: string) => {
  const blob = new Blob([data], { type: 'text/plain' })
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  a.click()
}

export const Homepage: FC = () => {
  const { loading, handleYTRequest } = useGetYTTranscription()
  const [ytUrl, setYTUrl] = useState<string>('')
  const [isInputInvalid, setIsInputInvalid] = useState<boolean>(false)

  const handleTranscribeDownload = async () => {
    try {
      const isYTUrlValid = isValidYoutubeVideoUrl(ytUrl)

      if (!isYTUrlValid) {
        setIsInputInvalid(true)
        return
      }

      setIsInputInvalid(false)

      const data = await handleYTRequest(ytUrl)

      if (data?.transcription) {
        saveToFile(
          data.transcription,
          `${data.videoId ? `${data.videoId}-` : ''}transcription.txt`,
        )
      } else {
        console.error('Missing transcription data')
      }
    } catch (error) {
      console.error(error)
    }
  }

  return (
    <main className="flex min-h-screen flex-col items-center justify-between p-24">
      <div className="flex flex-col items-center">
        <h1 className="text-4xl font-medium mb-8 text-gray-400">
          Youtube transcript downloader app
        </h1>

        <p className="text-gray-400 mb-7 max-w-96 text-center">
          Paste your Youtube video link in the input below to get its
          transcription. The transcription will be saved as a .txt file.
        </p>

        <Input
          className="mb-7"
          placeholder="Enter Youtube video link"
          value={ytUrl}
          onChange={e => setYTUrl(e.target.value)}
        />
        {isInputInvalid && (
          <p className="text-red-500 text-sm mb-7">
            Invalid Youtube video link
          </p>
        )}
        <Button
          disabled={loading || !ytUrl.length}
          onClick={handleTranscribeDownload}
        >
          Download transcript
        </Button>
      </div>
    </main>
  )
}
