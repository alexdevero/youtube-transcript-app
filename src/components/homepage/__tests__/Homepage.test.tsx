import { act, fireEvent, render } from '@testing-library/react'
import { Homepage } from '../Homepage'

describe('Homepage', () => {
  it('should render', () => {
    const { baseElement } = render(<Homepage />)
    expect(baseElement).toMatchSnapshot()
  })

  it('should render page title', () => {
    const { getByText } = render(<Homepage />)
    expect(getByText('Youtube transcript downloader app')).toBeInTheDocument()
  })

  it('should render page description', () => {
    const { getByText } = render(<Homepage />)
    expect(
      getByText(
        'Paste your Youtube video link in the input below to get its transcription. The transcription will be saved as a .txt file.',
      ),
    ).toBeInTheDocument()
  })

  it('should render input', () => {
    const { getByPlaceholderText } = render(<Homepage />)
    expect(getByPlaceholderText('Enter Youtube video link')).toBeInTheDocument()
  })

  it('should render submit button', () => {
    const { getByText } = render(<Homepage />)
    expect(getByText('Download transcript')).toBeInTheDocument()
  })

  it('should disable submit button when input is empty', () => {
    const { getByText } = render(<Homepage />)
    expect(getByText('Download transcript')).toBeDisabled()
  })

  it('should enable submit button when input is not empty', () => {
    const { getByText, getByPlaceholderText } = render(<Homepage />)
    const input = getByPlaceholderText('Enter Youtube video link')
    const button = getByText('Download transcript')

    act(() => {
      fireEvent.change(input, {
        target: { value: 'https://www.youtube.com/watch?v=123456' },
      })
    })

    expect(button).not.toBeDisabled()
  })

  it('should disable submit button when input is empty after being not empty', () => {
    const { getByText, getByPlaceholderText } = render(<Homepage />)
    const input = getByPlaceholderText('Enter Youtube video link')
    const button = getByText('Download transcript')

    act(() => {
      fireEvent.change(input, {
        target: { value: 'https://www.youtube.com/watch?v=123456' },
      })
    })

    act(() => {
      fireEvent.change(input, {
        target: { value: '' },
      })
    })

    expect(button).toBeDisabled()
  })

  it('should render error message when input is invalid', () => {
    const { getByText, getByPlaceholderText } = render(<Homepage />)
    const input = getByPlaceholderText('Enter Youtube video link')

    act(() => {
      fireEvent.change(input, {
        target: { value: 'invalid input' },
      })
    })

    act(() => {
      fireEvent.click(getByText('Download transcript'))
    })

    expect(getByText('Invalid Youtube video link')).toBeInTheDocument()
  })
})
