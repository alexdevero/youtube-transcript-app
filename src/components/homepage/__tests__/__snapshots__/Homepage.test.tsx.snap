// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Homepage > should render 1`] = `
<body>
  <div>
    <main
      class="flex min-h-screen flex-col items-center justify-between p-24"
    >
      <div
        class="flex flex-col items-center"
      >
        <h1
          class="text-4xl font-medium mb-8 text-gray-400"
        >
          Youtube transcript downloader app
        </h1>
        <p
          class="text-gray-400 mb-7 max-w-96 text-center"
        >
          Paste your Youtube video link in the input below to get its transcription. The transcription will be saved as a .txt file.
        </p>
        <input
          class="px-3 py-2 rounded-[4px] text-sm min-w-72 mb-7"
          placeholder="Enter Youtube video link"
          value=""
        />
        <button
          class="bg-blue-500 text-sm text-white px-4 py-2 rounded undefined opacity-50 cursor-not-allowed"
          disabled=""
        >
          Download transcript
        </button>
      </div>
    </main>
  </div>
</body>
`;
